### PR TITLE
gcode_macro: add "register_remote_method" endpoint

### PR DESCRIPTION
**Pull request by Arksine**
This adds the ability for webhooks clients to register methods that may be called from within a template.  The current implementation will allow for multiple clients to register the same method.  When invoked, the method is executed on all clients.

Signed-off-by:  Eric Callahan <arksine.code@gmail.com>

### CODE REVIEW

**Comment by KevinOConnor** [comment]
Interesting.  Thanks.  Some high-level comments:

- Since `register_remote_method` is effectively subscribing to asynchronous messages, I think it would be good to use the same format as the other subscription endpoints and take a `response_template` parameter.

- This functionality seems more related to the api server than gcode_macro, so it seems like it would be better located in webhooks.py.  (And have gcode_macro.py just add the `action_call_remote_method` hook.)

Separately, I'm hoping to let the v0.9.0 release settle for a bit, so this is something I'd look at merging late next week.

-Kevin

**Comment by Arksine** [comment]
Sounds good.  I'll move to webhooks and add the "response_template" as requested.

**Comment by Arksine** [comment]
I moved the "register_remote_method" endpoint to the webhooks module.  This implementation allows clients to re-register an endpoint in case they want to change the `response_template`.

File: klippy/webhooks.py [hunk 1]
```diff
@@ -320,6 +331,26 @@ def get_status(self, eventtime):
         state_message, state = self.printer.get_state_message()
         return {'state': state, 'state_message': state_message}
 
+    def call_remote_method(self, method, **kwargs):
+        if method not in self._remote_methods:
+            self.printer.lookup_object('gcode').respond_info(
+                "Remote method '%s' not registered" % (method))
```
Comment (KevinOConnor) [inline klippy/webhooks.py:337]: FWIW, I'm not sure `respond_info()` is a good way to report a problem.  I can't see regular users being able to find that info or make sense of it in the gcode output.

-Kevin

Comment (Arksine) [reply klippy/webhooks.py]: That makes sense.  I pushed some commits that raise a CommandError instead.  For `action_call_remote_method` I handle the error and log it, as I don't think we would want this particular error aborting a print.  If you think they look okay I'll squash them into the original commits.

File: klippy/webhooks.py [hunk 2]
```diff
@@ -320,6 +331,24 @@ def get_status(self, eventtime):
         state_message, state = self.printer.get_state_message()
         return {'state': state, 'state_message': state_message}
 
+    def call_remote_method(self, method, **kwargs):
+        if method not in self._remote_methods:
+            raise homing.CommandError(
```
Comment (KevinOConnor) [inline klippy/webhooks.py:336]: Very minor, but it's preferable to `raise self.printer.command_error()` than to use the direct `homing` reference.

In any case, the code as a whole looks okay to me.

Thanks,
-Kevin

Comment (Arksine) [reply klippy/webhooks.py]: I fixed this and squashed everything into 3 commits.  Thanks!

**Comment by KevinOConnor** [comment]
Thanks.

-Kevin

**Pull request merged**